### PR TITLE
Hide hero shadow in castle

### PR DESCRIPTION
**Pull request by vasilenkoalexey**
Hide hero shadow when hero in castle (like in OG)

### CODE REVIEW

File: src/fheroes2/heroes/heroes_move.cpp [hunk 1]
```diff
@@ -475,6 +475,22 @@ void Heroes::Redraw( fheroes2::Image & dst, int32_t dx, int32_t dy, const Rect &
         }
     }
 
+    const Maps::Tiles & tile = world.GetTiles( center.x, center.y );
```
Comment (ihhub) [inline src/fheroes2/heroes/heroes_move.cpp:478]: Let's call it `heroTile` instead other if-bodies can have the same name.

File: src/fheroes2/heroes/heroes_move.cpp [hunk 2]
```diff
@@ -475,6 +475,22 @@ void Heroes::Redraw( fheroes2::Image & dst, int32_t dx, int32_t dy, const Rect &
...
+
+    if ( withShadow ) {
+        if ( tile.GetObject( false ) == MP2::OBJ_CASTLE && ( sprite_index > 44 || ( sprite_index % 9 ) < 5 ) ) {
```
Comment (ihhub) [inline src/fheroes2/heroes/heroes_move.cpp:481]: Checking sprite index of a hero might be a bad idea. Please check with dimension door and town portal spells.

**Review by ihhub — CHANGES_REQUESTED** [review]
Hi @vasilenkoalexey , I left 2 comments. Could you please take a look?

File: src/fheroes2/heroes/heroes_move.cpp (hunk 2 above)
Comment (vasilenkoalexey) [reply src/fheroes2/heroes/heroes_move.cpp:481]: This is the definition that the hero is moving and he is somewhere in the middle between the bottom tile and top castle tile. I dont known how town portal spells can help here :(

Comment (ihhub) [reply src/fheroes2/heroes/heroes_move.cpp]: For example, I used Town Portal spell to teleport a hero to a castle. As you could see the location of the hero could be any:
image

Comment (vasilenkoalexey) [reply src/fheroes2/heroes/heroes_move.cpp]: > As you could see the location of the hero could be any:

Do you mean direction? I dont see any problem here - hero on castle, and there is no hero shadow. `sprite_index > 44` mean hero turns in place without moving (we should not render hero shadow when he just turn in castle). `( sprite_index % 9 ) < 5` mean hero goes out of the castle but has passed less than "half way" to the bottom tile. And we start to draw hero shadow only if he has passed more than "half way". At least I see this kind of logic in  the OG... I don't understand why we can't use sprite index here, and how Town Portal Spell can help implement this. Could you explain in more detail what you mean?

Comment (ihhub) [reply src/fheroes2/heroes/heroes_move.cpp]: I meant that while using Town Portal spell we don't the direction of a hero leading to a false condition for your code which later shows a shadow of the hero.

Give me please some time to come up with a solution to cover all cases.

Comment (vasilenkoalexey) [reply src/fheroes2/heroes/heroes_move.cpp]: > Give me please some time to come up with a solution to cover all cases.

Shure. But for all hero directions `sprite_index % 9` always is zero (if hero not moving or turning). I decided to check "with  eyes" for all Town Portal directions and did not see the hero shadow
hero

**Comment by ihhub** [comment]
Hi @vasilenkoalexey ,

actually in the original game hero has a shadow while in a castle:
image

**Comment by vasilenkoalexey** [comment]
> actually in the original game hero has a shadow while in a castle:

Hmm... You right, but castle drawn over shadow... So i will cover this case in fututre PR with drawing the hero and the shadow in different loops.

**Pull request closed without merging**